src/DsaLocal.cc

```diff
@@ -1084,6 +1087,46 @@ void LocalAnalysis::runOnFunction(Function &F, Graph &g) {
   );
   // clang-format on
 
+  /// Mark incompelete nodes after the pass
+  /// based on the code in remove_dead
+  /// TODO: doing it inside remove_dead maybe more efficient
+  std::set<const Node *> reachable;
+  // --- collect all nodes referenced by formal parameters
+  for (auto &kv : llvm::make_range(g.formal_begin(), g.formal_end())) {
+    const Cell *C = kv.second.get();
+    if (C->isNull())
+      continue;
+    if (reachable.insert(C->getNode()).second) {
+      LOG("dsa-callsite", errs() << "\treachable formal parameter nodes " << C->getNode() << "\n";);
+      C->getNode()->setIncomplete(true);
+    }
+  }
+
+  // --- collect all nodes referenced by return parameters
+  for (auto &kv : llvm::make_range(g.return_begin(), g.return_end())) {
+    const Cell *C = kv.second.get();
+    if (C->isNull())
+      continue;
+    if (reachable.insert(C->getNode()).second) {
+      LOG("dsa-dead", errs() << "\treachable return parameter nodes " << C->getNode() << "\n";);
+      C->getNode()->setIncomplete(true);
+    }
+  }
+
+  // --- collect all nodes referenced by global parameters
+  if(F.getName()!="main"){
+    for (auto &kv : llvm::make_range(g.globals_begin(), g.globals_end())){
```

Comment:
Again, need to go deeper. There should be code for finding what is reachable from a node already. 
This is used, for example, in bottom-up to move only nodes that might change from callee to caller.